### PR TITLE
U/fjammes/dm 3985

### PR DESCRIPTION
**Pull request by fjammes**

### CODE REVIEW

File: admin/tools/docker/integration-tests/localhost/env.example.sh [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# VERSION can be ticket branch but with _ instead of /
```
Comment (jhoblitt) [inline admin/tools/docker/integration-tests/localhost/env.example.sh]: This file was added to the repo without execute permissions -- was that intentional.  Ie., it should only ever be sourced?

Comment (jhoblitt) [inline admin/tools/docker/integration-tests/localhost/env.example.sh]: No shebang line.

File: admin/tools/docker/integration-tests/localhost/pull.sh [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#!/bin/sh
+
+# Download Docker container related to Qserv integration tests
+
+# @author  Fabrice Jammes, IN2P3/SLAC
+
+. ./env.sh
```
Comment (jhoblitt) [inline admin/tools/docker/integration-tests/localhost/pull.sh]: Is the intent that `env.example.sh` is renamed by the user to `env.sh`?  If so, a comment to that affect might be useful.

File: admin/tools/docker/integration-tests/localhost/pull.sh [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+
+docker pull ${MASTER_IMAGE}
+docker pull ${WORKER_IMAGE}
```
Comment (jhoblitt) [inline admin/tools/docker/integration-tests/localhost/pull.sh]: Missing quotes

File: admin/tools/docker/integration-tests/localhost/run-multinode-tests.sh [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the LSST License Statement and
+# the GNU General Public License along with this program.  If not,
+# see <http://www.lsstcorp.org/LegalNotices/>.
```
Comment (jhoblitt) [inline admin/tools/docker/integration-tests/localhost/run-multinode-tests.sh]: The license boilerplate should not be nessicary.

File: admin/tools/docker/integration-tests/localhost/run-multinode-tests.sh [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+usage() {
+  cat << EOD
+
+  Usage: $(basename "$0") [options]
+
+  Available options:
+    -h          this message
+    -C          Rebuild the images from scratch
+
+  Launch Qserv integration tests on one Docker host
```
Comment (jhoblitt) [inline admin/tools/docker/integration-tests/localhost/run-multinode-tests.sh]: trailing whitespace

File: admin/tools/docker/integration-tests/localhost/run-multinode-tests.sh [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+  Launch Qserv integration tests on one Docker host 
+
+EOD
+}
+
+# get the options
+while getopts hC c ; do
+    case $c in
+            h) usage ; exit 0 ;;
```
Comment (jhoblitt) [inline admin/tools/docker/integration-tests/localhost/run-multinode-tests.sh]: I think there is an extra level of indention here.

Comment (jhoblitt) [inline admin/tools/docker/integration-tests/localhost/run-multinode-tests.sh]: why is `-h` a normal exit status while `-?` is not?  Could these be combined into a single case as `h | \? )`?

File: admin/tools/docker/integration-tests/localhost/run-multinode-tests.sh [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+            C) CACHE_OPT="--no-cache=true" ;;
```
Comment (jhoblitt) [inline admin/tools/docker/integration-tests/localhost/run-multinode-tests.sh]: Where is `CACHE_OPT` used?

File: admin/tools/docker/integration-tests/localhost/run-multinode-tests.sh [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    esac
+done
+shift "$((OPTIND-1))"
+
+if [ $# -ne 0 ] ; then
+    usage
+    exit 2
+fi
+
+# Start dynamic DNS for local Docker containers
```
Comment (jhoblitt) [inline admin/tools/docker/integration-tests/localhost/run-multinode-tests.sh]: trailing whitespace

File: admin/tools/docker/integration-tests/localhost/run-multinode-tests.sh [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+if [ $# -ne 0 ] ; then
+    usage
+    exit 2
+fi
+
+# Start dynamic DNS for local Docker containers 
+docker rm -f docker_spy || echo "No existing container for docker_spy"
+docker run --name docker_spy --detach=true -e "DNS_DOMAIN=$DNS_DOMAIN" -v /var/run/docker.sock:/var/run/docker.sock iverberk/docker-spy
+
+DNS_IP=$(docker inspect -f '{{ .NetworkSettings.IPAddress }}' docker_spy)
```
Comment (jhoblitt) [inline admin/tools/docker/integration-tests/localhost/run-multinode-tests.sh]: This is clever. :+1:

File: admin/tools/docker/integration-tests/localhost/run-multinode-tests.sh [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+
+docker rm -f "$MASTER" || echo "No existing container for $MASTER"
+docker run --detach=true --dns="$DNS_IP" --dns-search="$DNS_DOMAIN" --name "$MASTER" -h "${MASTER}" "$MASTER_IMAGE"
```
Comment (jhoblitt) [inline admin/tools/docker/integration-tests/localhost/run-multinode-tests.sh]: trailing whitespace

File: admin/tools/docker/integration-tests/localhost/run-multinode-tests.sh [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+DNS_IP=$(docker inspect -f '{{ .NetworkSettings.IPAddress }}' docker_spy)
+
+docker rm -f "$MASTER" || echo "No existing container for $MASTER"
+docker run --detach=true --dns="$DNS_IP" --dns-search="$DNS_DOMAIN" --name "$MASTER" -h "${MASTER}" "$MASTER_IMAGE" 
+
+for i in $WORKERS;
+do
+    docker rm -f "$i" || echo "No existing container for $i"
+    docker run --detach=true --dns="$DNS_IP" --dns-search="$DNS_DOMAIN" --name "$i" -h "${i}"  "$WORKER_IMAGE"
+done
```
Comment (jhoblitt) [inline admin/tools/docker/integration-tests/localhost/run-multinode-tests.sh]: trailing whitespace

File: doc/source/HOW-TO/docker-multinode-test-localhost.rst [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+******************
+Run multinode test
+******************
+
+On the workstation *myhost.domain.org*, clone Qserv code and go to directory containing tests scripts.
+
+.. code-block:: bash
+
+   git clone git@github.com:lsst/qserv.git
+   cd ${SRC_DIR}/qserv/admin/tools/docker/integration-tests/localhost
```
Comment (jhoblitt) [inline doc/source/HOW-TO/docker-multinode-test-localhost.rst]: I think `${SRC_DIR}/` can be removed from this example.

File: admin/tools/docker/integration-tests/localhost/run-multinode-tests.sh [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+shift "$((OPTIND-1))"
+
+if [ $# -ne 0 ] ; then
+    usage
+    exit 2
+fi
+
+# Start dynamic DNS for local Docker containers 
+docker rm -f docker_spy || echo "No existing container for docker_spy"
+docker run --name docker_spy --detach=true -e "DNS_DOMAIN=$DNS_DOMAIN" -v /var/run/docker.sock:/var/run/docker.sock iverberk/docker-spy
```
Comment (jhoblitt) [inline admin/tools/docker/integration-tests/localhost/run-multinode-tests.sh]: Passing the docker sock down into a container is a fairly risky thing to do.  I think this warrants a comment noting that this is a security concern.

File: admin/tools/docker/integration-tests/localhost/.gitignore [hunk 14]
```diff
@@ -0,0 +1,2 @@
+/env.sh
+/nodes.css
```
Comment (jhoblitt) [inline admin/tools/docker/integration-tests/localhost/.gitignore]: Is `nodes.css` still used?

File: admin/tools/docker/integration-tests/localhost/env.example.sh (hunk 1 above)
Comment (fjammes) [inline admin/tools/docker/integration-tests/localhost/env.example.sh]: Yes file should only be sourced, so shebang isn't required. Comment added about this

File: admin/tools/docker/integration-tests/localhost/pull.sh (hunk 2 above)
Comment (fjammes) [inline admin/tools/docker/integration-tests/localhost/pull.sh]: :+1:

File: admin/tools/docker/integration-tests/localhost/run-multinode-tests.sh (hunk 6 above)
Comment (fjammes) [inline admin/tools/docker/integration-tests/localhost/run-multinode-tests.sh]: I think current behaviour is normal, for example try "python -h" and "pŷthon -azerty"

File: admin/tools/docker/integration-tests/localhost/run-multinode-tests.sh (hunk 7 above)
Comment (fjammes) [inline admin/tools/docker/integration-tests/localhost/run-multinode-tests.sh]: :+1: removed

File: admin/tools/docker/integration-tests/localhost/run-multinode-tests.sh (hunk 9 above)
Comment (fjammes) [inline admin/tools/docker/integration-tests/localhost/run-multinode-tests.sh]: thanks :-).

File: admin/tools/docker/integration-tests/localhost/run-multinode-tests.sh (hunk 13 above)
Comment (fjammes) [inline admin/tools/docker/integration-tests/localhost/run-multinode-tests.sh]: :+1:

File: admin/tools/docker/integration-tests/localhost/.gitignore (hunk 14 above)
Comment (fjammes) [inline admin/tools/docker/integration-tests/localhost/.gitignore]: Yes, it's no more used, thanks :-)

**Pull request merged**